controller/backup_controller.go

```diff
@@ -279,29 +287,97 @@ func (bc *BackupController) reconcile(backupName string) (err error) {
-		// Initialize a backup target client
-		backupTargetClient, err := getBackupTargetClient(bc.ds, backupTarget)
-		if err != nil {
-			log.WithError(err).Error("Error init backup target client")
-			return nil // Ignore error to prevent enqueue
-		}
+	if backup.Spec.SnapshotName != "" {
+		switch backup.Status.State {
+		case types.BackupStateNew:
+			// Initialize a backup target client
+			backupTargetClient, err := getBackupTargetClient(bc.ds, backupTarget)
+			if err != nil {
+				log.WithError(err).Error("Error init backup target client")
+				return nil // Ignore error to prevent enqueue
+			}
 
-		// Initialize a engine client
-		engine, err := bc.ds.GetVolumeCurrentEngine(backupVolumeName)
-		if err != nil {
-			return err
-		}
-		engineCollection := &engineapi.EngineCollection{}
-		engineClient, err := GetClientForEngine(engine, engineCollection, engine.Status.CurrentImage)
-		if err != nil {
-			return err
-		}
+			// Initialize a engine client
+			engine, err := bc.ds.GetVolumeCurrentEngine(backupVolumeName)
+			if err != nil {
+				return err
+			}
+			engineCollection := &engineapi.EngineCollection{}
+			engineClient, err := GetClientForEngine(engine, engineCollection, engine.Status.CurrentImage)
+			if err != nil {
+				return err
+			}
 
-		if err := bc.backupCreation(log, engineClient, backupTargetClient.URL, backupTargetClient.Credential, backup); err != nil {
-			return err
+			if err := bc.backupCreation(log, engineClient, backupTargetClient.URL, backupTargetClient.Credential, backup); err != nil {
+				return err
+			}
+
+			// Monitor snapshot backup progress
+			bc.enableBackupMonitor(backup, engineClient)
+			return nil
+		case types.BackupStateInProgress:
+			bm := bc.hasMonitor(backup.Name)
+			if bm == nil || bm.engineBackupStatus == nil {
+				return nil
+			}
+
+			backup.Status.Progress = bm.engineBackupStatus.Progress
+			backup.Status.URL = bm.engineBackupStatus.BackupURL
+			backup.Status.Error = bm.engineBackupStatus.Error
+			backup.Status.SnapshotName = bm.engineBackupStatus.SnapshotName
+			backup.Status.ReplicaAddress = bm.engineBackupStatus.ReplicaAddress
+			// https://github.com/longhorn/longhorn-engine/blob/9da3616/pkg/replica/backup.go#L20-L22
+			switch bm.engineBackupStatus.State {
+			case "in_progress":
+				backup.Status.State = types.BackupStateInProgress
+			case "complete":
+				backup.Status.State = types.BackupStateCompleted
+			case "error":
+				backup.Status.State = types.BackupStateError
+			}
+
+			// log+event when backup state changes
+			if existingBackup.Status.State != backup.Status.State {
+				// Get the volume CR
+				volume, err := bc.ds.GetVolume(backupVolumeName)
+				if err != nil {
+					return err
+				}
+				bc.eventLog(nil, backup.Status.State, backup, volume)
+			}
+
+			// state changes from in progress to complete
+			if backup.Status.State == types.BackupStateCompleted {
+				// disable backup monitor
+				bc.disableBackupMonitor(backup)
```

Comment:
Putting this line here implicitly means the monitor stop is an edge-triggered operation and it happens only when the current backup state is updated from `in_progress` to `completed`. 
IMO, this should be a level-triggered operation instead: As long as the current backup reaches an end state (completed, error, not sure if unknown should be included), the monitor can be stopped.